lib/instrumentation/grpc-js/grpc.js

```diff
@@ -6,80 +6,206 @@
+        segment.end()
 
-          const protocol = 'grpc'
+        originalListener && originalListener.onReceiveStatus(status)
+      }
 
-          const url = `${protocol}://${authorityName}${this.methodName}`
+      args[1] = nrListener
 
-          segment.addAttribute('http.url', url)
-          segment.addAttribute('http.method', this.methodName)
+      return original.apply(this, args)
+    }
+  }
+}
 
-          segment.end()
+/**
+ * Instruments the grpc-js server by intercepting the moment when
+ * server methods are registered from the method implementations
+ * provided to grpc-js. This handles all four types of server
+ * invocations: unary, client-streaming, server-streaming, and
+ * bidirectional streaming.
+ *
+ * @param {object} shim the web shim to instrument with
+ * @param {Function} original the original function
+ * @returns {Function} the instrumented function
+ */
+function wrapRegister(shim, original) {
+  const constants = shim.require('./build/src/constants')
+
+  return function wrappedRegister() {
+    const args = shim.argsToArray.apply(shim, arguments)
+
+    const name = args[0]
+    const handler = args[1]
+    const type = args[args.length - 1]
+    // We need the server for the host and port, but depending on how
+    // the user calls server functions, the connexion might not be
+    // ready yet. It will be ready below by the time the registered
+    // handler is invoked.
+    const server = this
+
+    if (this.handlers.has(name)) {
+      shim.logger.debug(`gRPC handler for ${name} won't be added so not instrumenting`)
+      return original.apply(this, arguments)
+    }
 
-          originalListener && originalListener.onReceiveStatus(status)
-        }
+    args[1] = shim.bindCreateTransaction(createTransaction, { type: shim.WEB })
 
-        args[1] = nrListener
+    return original.apply(this, args)
 
-        return original.apply(this, args)
-      }
+    function createTransaction(stream) {
+      const parent = shim.getActiveSegment()
+      const transaction = parent.transaction
+      const address = server.http2ServerList[0].channelzRef.name
+
+      // Create the transaction segment using the request URL for now. Once a
+      // better name can be determined this segment will be renamed to that.
+      const segment = shim.createSegment(name, recordHttp)
+      segment.start()
+
+      transaction.type = 'web'
+      transaction.baseSegment = segment
+      transaction.url = name
+
+      transaction.trace.attributes.addAttribute(DESTINATION, 'request.method', transaction.url)
+
+      transaction.trace.attributes.addAttribute(
+        DESTINATION,
+        'request.uri',
+        `grpc://${address}${transaction.url}`
+      )
+
+      shim.setTransactionUri(transaction.url)
+
+      // This attribute isn't required by the spec, but it seems useful to have
+      transaction.trace.attributes.addAttribute(DESTINATION, 'grpc.type', type)
+
+      const metadata = stream.metadata
```

Comment:
feels like the DT logic should be extracted to its own function.  for someone unfamiliar it's quite a few steps to extract and accept "headers"

Reply:
Okay, does that look better?

Reply:
yea for sure. nice work 👏🏻